### ISSUE TITLE
Some BPG IPv6 withdrawal events no longer register. RPKI ROA events and IPv4 withdrawal events are detected correctly.

### ISSUE THREAD

**Issue by mfld-pub** [nttgin/BGPalerter#993]
**Describe the bug**
Two prod instances at two different vantage points are configured to report on withdrawal events, hijacks, unexpected prefixes from the ASN. Since upgrading, these events do not register. Notifications are not triggered and nothing is in reports.log

**Provide an example**
As the prod instances have convoluted yml that have survived years of version upgrade I deployed a brand new Debian 11 instance and started from scratch for testing.

- AS212996 is announcing 2a0c:9a40:8d60::/44
- Install the [latest version ](https://github.com/nttgin/BGPalerter/releases/download/v1.31.0/bgpalerter-linux-x64) on a new Debian 11 instance that has dual stack connectivity and no egress firewall.
- Run the onboarding. `chmod +x bgpalerter-linux-x64 && ./bgpalerter-linux-x64`
- Wait for it to confirm it is running:

![Screenshot 2022-12-21 164712](https://user-images.githubusercontent.com/66289900/208959406-37050dac-d88c-455d-be2b-a581d4c5da36.png)

- In ./logs I monitor errors and reports `tail -f reports.log` ,  `tail -f errors.log`
- Open a RISlive connection and subscribe for updates filtering for the ASN
```
// Received at 16:52:19
{
    "type": "ris_subscribe_ok",
    "data": {"subscription":{"host":"rrc04.ripe.net","path":"212996$"},"socketOptions":{"includeRaw":false,"acknowledge":true}}
}
```

- Withdraw the prefix and wait.
- See the withdrawal in RISlive
- See no activity in reports.log. error.log just shows subscription to RIS when I started the process:
```
2022-12-21T16:46:15+00:00 info: ris connector connected (instance:e147e27a-7a8d-408b-9255-e07299794665 connection:5204c8fb-0ada-45ad-90bb-dca42549ae81)
2022-12-21T16:46:15+00:00 info: Subscribed to monitored resources
2022-12-21T16:46:15+00:00 info: Subscribed to monitored resources
2022-12-21T16:46:15+00:00 info: Subscribed to beacons
2022-12-21T16:46:15+00:00 info: Subscribed to beacons
2022-12-21T16:46:15+00:00 info: Prefix rules reloaded
```

What I have done to troubleshoot:

- Deploy new on dev instance
- Set thresholdMinPeers: 5 for withdrawal-detection in config.xml 
```
  - file: monitorVisibility
    channel: visibility
    name: withdrawal-detection
    params:
      thresholdMinPeers: 5
```

**Expected behavior**
withdrawal-detection should have triggered and logged to reports.log once I could see the prefix gone in RISlive

**Are you using the binary or the source code?**
linux-64 binary

**Your information**
Michael Felden, AS212996

**Comment by mfld-pub** [comment]
FWIW if I empty out the auto generated prefixes.yml to just monitor the ASN, it does detect announcements so its clearly talking to the RIS and able to write to reports.log

```
options:
  monitorASns:
    '212996':
      group: noc
      upstreams:
        - 20473
        - 34927
      downstreams: []
  generate:
    exclude: []
    excludeDelegated: true
    monitoredASes:
      - '212996'
    historical: false
    group: noc
    asnList:
      - '212996'
```

`2022-12-21T17:12:51+00:00 verbose: AS212996 is announcing 2a0c:9a40:8d60::/44 but this prefix is not in the configured list of announced prefixes` 

Just withdrawal events (and maybe other events) of known prefixes seem to be ignored.

**Comment by massimocandela** [comment]
Hi @mfld-pub, can you provide me with a utc timestamp of when you attempted the withdrawal of the prefix? Please, provide me with an attempt older than 2 hours if available.

**Comment by massimocandela** [comment]
Also, if you can attach the prefixes.yml you used for the experiment and the config.yml stripped out of passwords and other chat configuration. Feel free to send it by email.

**Comment by mfld-pub** [comment]
> Also, if you can attach the prefixes.yml you used for the experiment and the config.yml stripped out of passwords and other chat configuration. Feel free to send it by email.

Sure thing. The config.xml is not sensitive as I did an install from scratch for this and there is only reporting to log file enabled on this dev instance. have attached them here in D.zip

I did uncomment 

```
  - file: monitorNewPrefix
    channel: newprefix
    name: prefix-detection
    params:
      thresholdMinPeers: 3
```

and I did set thresholdMinPeers: 3 for channel: visibility to make it super sensitive for testing purposes.

```
  - file: monitorVisibility
    channel: visibility
    name: withdrawal-detection
    params:
      thresholdMinPeers: 3
```

The prefixes.yml are the auto generated kind you get from initial run. The test prefix is in there.

An example where I withdrew test prefix 2a0c:9a40:8d60::/44 can be seen in RIPEstat (in UTC):

```
From 2022-12-21 14:53 to 2022-12-21 14:54

2a0c:9a40:8d60::/44	W		[152.89.170.244](https://stat.ripe.net/app/launchpad/152.89.170.244)	25
2a0c:9a40:8d60::/44	W		[2a05:4140:3::476](https://stat.ripe.net/app/launchpad/2a05:4140:3::476)	25
```

The two config files:

[D.zip](https://github.com/nttgin/BGPalerter/files/10280248/D.zip)

**Comment by massimocandela** [comment]
Hi @mfld-pub,

After some tests I believe that nothing abnormal is going on. I also checked several of my production instances and they have been reporting withdrawals consistently. In one of this instance I had 22 withdrawal alerts (different prefixes) in the last 24 hours. However, this piece of software is really important and I don't want to discard anything so I propose below another test for you, in addition to reporting the results of the tests I did.

If I look at the withdrawals around the time you reported (14:30 UTC to 15:16 UTC), they are visible by only 1 RIS peer [1]. By default BGPalerter sets that withdrawal threshold to at least 40 peers (a really small amount compared to the total amount of RIS peers, but big enough to avoid noise). Even with your initial test configuration of 5 peers, and with the one you reported in the last message of 3 peers, nothing would have been triggered.

Even if we take a wider window, from 10:00 to 18:00 UTC [2][3], we have withdrawals visible by only 6 peers and spread across the 8 hours time frame. The cooling period would have correctly prevented the alert as common/irrelevant unrelated withdrawals. Furthermore, it looks the withdrawals are followed by a new announcement in 2 to 4 minutes (in addition to the other examples, look at [4]). For BGPalerter to trigger an alert, the visibility should be lost for at least X peers (completely, no path changes, see case [5] which may be similar to yours). 
To validate this, I imported the updates from the BGPlay api in BGPalerter. With thresholdMinPeers:1 I can see the withdrawals and I receive the alert. I can also see them if I increase the thresholdMinPeers and I remove the cooling threshold. This config is just for test, it makes for a useless monitoring. Alerting on withdrawals visible from a few peers and for really short period of time would produce a lot of useless noise. 

Maybe, what you are interested in your production configuration is to reduce the notificationIntervalSeconds to allow for repeated visibility notification about the same prefix in the same day (in case you are monitoring some kind of beacon).

Further test for you:
- use default configuration and change the thresholdMinPeers to 1 (due to the low-visibility seen above)
- set notificationIntervalSeconds: 1 
- remove ".cache/" at every restart of your test
- use a prefix that is monitored in prefixes.yml for AS212996. The prefix must be initially visible (let's say, announced by at least 15 minutes).
- withdraw the prefix and do not announce it back for at least 15 min (random amount, big enough).
- check if you receive a log in reports.log
- check in the BGPlay api below if you are able to see events type: "W". To do that you just need to manipulate the timestamp input parameters in the url. If you are counting the peers, you need to use unique source_id IPs. You may need to wait before to be able to see the data in the BGPlay api.


[1] https://stat.ripe.net/data/bgplay/data.json?endtime=1671635805&resource=2a0c%3A9a40%3A8d60%3A%3A%2F44&rrcs=0%2C1%2C2%2C3%2C4%2C5%2C6%2C7%2C10%2C11%2C12%2C13%2C14%2C15%2C16%2C18%2C19%2C20%2C21&starttime=1671633180&unix_timestamps=TRUE

[2] https://stat.ripe.net/data/bgplay/data.json?endtime=1671667199&resource=2a0c%3A9a40%3A8d60%3A%3A%2F44&rrcs=0%2C1%2C2%2C3%2C4%2C5%2C6%2C7%2C10%2C11%2C12%2C13%2C14%2C15%2C16%2C18%2C19%2C20%2C21&starttime=1671616800&unix_timestamps=TRUE

[3] https://stat.ripe.net/ui2013/widget/bgplay#w.resource=2a0c:9a40:8d60::/44&w.ignoreReannouncements=false&w.starttime=1671616800&w.endtime=1671645600&w.rrcs=0,1,2,3,4,5,6,7,10,11,12,13,14,15,16,18,19,20,21&w.instant=null&w.type=bgp

[4] https://stat.ripe.net/ui2013/widget/bgplay#w.resource=2a0c:9a40:8d60::/44&w.ignoreReannouncements=false&w.starttime=1671639360&w.endtime=1671639838&w.rrcs=0,1,2,5,6,7,10,11,13,14,15,16,18,20&w.instant=null&w.type=bgp

[5] https://github.com/nttgin/BGPalerter/issues/946

**Comment by massimocandela** [comment]
Reopening this issue.

Yesterday, with @mfld-pub we spent most of our day with him announcing and withdrawing stuff and me detecting it.
He had a hunch that in the past weeks the number of alerts he received about disappearing prefixes was lower compared to previous times. He provided me with an example, different from the one reported above, where indeed the count of the peers should have been able to trigger the alert. 

This time I was able to find the issue and fix it. A patch will be released soon.

Summary: The issue was impacting the sensitivity of withdrawal detection (mostly low-visibility prefixes), for the past ~1.5 months, only IPv6.

Details: A filter was wrongly differentiating between short and long representation of ipv6 addresses during withdrawal detection. This was arbitrarily impacting the count of the peers. If the count of the peers was impacted, a disappearing low-visibility prefix could have been excluded by the thresholdMinPeers. This started to happen when recently some improvements where deployed to the RIS architecture (before ipv6 were all in the short format; however, this is not an issue with RIS). The issue is not reproducible with historical data, which uses only v6 in short format (this made the debugging more difficult). 

A big thank to @mfld-pub!!
